### DISCUSSION TITLE
Excessive view re-rendering

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#1435] [Q&A] [answered]
Hello 😄

Our app is having an issue where views are rendered excessively often. After reviewing the [performance document](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance), we started applying the recommended scoping strategy (except that we're explicitly creating the scoped stores rather than use `WithViewStore`, for reasons that aren't relevant to this discussion) but we noticed that views whose states have not changed are still re-rendered.

We've built a really simple toy project that shows the issue, and the repo is publicly visible [here](https://github.com/wtruppel/ScopingToyProj). The toy project builds an app that presents the following UI:

![screenshot](https://user-images.githubusercontent.com/91148917/193624087-3d63cf2e-030a-4e66-9596-705575a22900.png)

When the user taps the "MIcrophone is on" toggle, only ChildSceneA's view gets re-rendered, which is the correct behaviour. So far, so good.

But if the user taps either toggle for "User is logged in", the ChildSceneA's view _still_ gets re-rendered, even though the state of SceneChildA has not changed.

I understand that the root scene's state has changed and, so, its view needs re-rendering but I'd still expect any subviews whose states have not changed to just stay put.

Any help would be greatly appreciated. Many thanks in advance.
Cheers!

**Comment by tgrapperon** [comment, marked as answer]
Hey @wtruppel, thanks for the toy project!
This is the `ViewStore` you create in `ChildSceneA`. Its instance changes when `ChildSceneA` initializes, and this forces a re-rendering of `body`. As you are iOS 14+, you can upgrade your `@ObservedObject` to `@StateObject` (which evaluate its `wrappedValue` only once with an `autoclosure`) to fix the issue:
```swift
@StateObject private var viewStore: ViewStore<State, Action>
public init(store: Store<State, Action>) {
  self._viewStore = .init(wrappedValue: ViewStore(store))
}
```
So when you toggle in `ChildSceneB`, you invalidate the `ViewStore` in `RootScene` which makes `RootScene` reruns its `body` and regenerate `ChildSceneA`, and then a create new `ViewStore` instance, and then a `ChildSceneA` body because `ChildSceneA` dependencies have then changed. You would have the same issue if you would scope the root store in `RootScene.body`, but I saw that you effectively avoided doing that by scoping in the `init` (which is probably not a good idea in general, albeit not problematic in this root position).

Please also note that you can use `_ = Self._printChanges()` in your `body`s in iOS 15+ to get a better view about what caused the evaluation.

**Reply by ghost** [reply]
Hi @tgrapperon. Thank you so much for replying. 

Yes, that solves it! Thank you!! Why did you say this, though?

> scoping in the init (which is probably not a good idea in general...)

**Reply by tgrapperon** [reply]
Because it takes time, and `View`'s `init` is assumed by SwiftUI to be very lean, as it can run an arbitrary number of times. This root view is likely initialized only once, but if it was a leaf view, the two store instances you would create there and keep as properties would cascade into a `body` run (because SwiftUI would detect that the `View`'s dependencies are different). 

With the current version of TCA, `store.scope` always returns a new instance, and thus always invalidates the view's it's passed to as an argument (provided the parent's `body` where the scope happens runs). The best you can do to limit this effect is thus to prevent the parent's `body` to run, hence to properly scope its `ViewStore`'s `State`. Optimally `store.scope` and `ViewStore(store)` would return the same instance each time, allowing shorter invalidation orbits. There are some experiments on the matter. 

In the meantime, scoping in `body` would make your re-render issue unavoidable I'm afraid. This is usually not the most problematic one, and if you spot a bottleneck, you can maybe fall back to scoping in the `init` if it ever helps your case. The `@StateObject` trick won't work anymore, as you end up capturing different `Store` instances in the `autoclosure`. This will in turn change the capture list of this closure, and this will warrant a new body evaluation (but the closure won't be re-evaluated).
 
Scoping in the instance in the `init` is something that is effective only as long as the view is not recreated, like it seems to be the case for the app's root view. If it is possible that it can offer local workarounds, I wouldn't advise to do like this in the general picture. If you encounter a situation where it really makes a difference in terms of performances, please share with us, as it would then make a good test case for stable scoped and `ViewStore` instances.

You can learn more about SwiftUI's invalidation process and dependencies in the WWDC 21 session ["Demystify SwiftUI"](https://developer.apple.com/videos/play/wwdc2021/10022/).

**Reply by tgrapperon** [reply]
[From this year's SwiftUI's WWDC Lounge](https://wwdclounges.com/wwdc22/swiftui-lounge.html#why-does-navigationlink-pre-render-the-destination-when-its-being-rendered-i-am-loading-heart-rate-data-in-my-destination-view-and-this-slows-down-rendering-significantly-whats-a-better-pattern-to-prevent-accidental-pre-loading-of-data):
> In general we discourage to do any expensive work in the init of a view. SwiftUI expects the creation of views to be cheap.

Of course `store.scope` is relatively cheaper than a network request, so your mileage may vary.

**Reply by ghost** [reply]
Thank you, @tgrapperon, for the extra explanation. 👍 

Meanwhile, a colleague noticed that not using any extra stores or view stores in the views, like this
```swift
extension RootScene {

    public struct View: SwiftUI.View {

        private let store: Store<State, Action>

        public init(store: Store<State, Action>) {
            self.store = store
        }

        public var body: some SwiftUI.View {
            print("RootScene.View rendered")
            return GroupBox(label: Text("RootScene") ) {
                VStack {

                    WithViewStore(store, observe: \.userIsLoggedIn) { viewStore in
                        Toggle(
                            "User is logged in",
                            isOn: viewStore.binding(send: .userIsLoggedInToggleTapped)
                        )
                    }

                    ChildSceneA.View(store: store.scope(
                        state: \.childSceneA,
                        action: Action.childSceneA
                    ))
                    .padding(15)

                    ChildSceneB.View(store: store.scope(
                        state: \.childSceneB,
                        action: Action.childSceneB
                    ))
                    .padding(15)

                }
            }
        }

    }

}
```
(and similarly for the other views, full code [here](https://github.com/wtruppel/ScopingToyProj/tree/withViewStore)) not only appears to do the right thing but also no print statement is ever executed, other than when the app launches.

That looks very mysterious to us because it shows that none of the body methods are being executed (since the "View rendered" lines are not printing), yet all the views that need to be re-rendered are re-rendered, and only those.

If you could shed some light on that mystery, we'd be very grateful.

**Reply by tgrapperon** [reply]
Indeed, this is the magic of SwiftUI invalidation coupled with proper `ViewState` scoping and `WithViewStore` convenient compartmenting.

In your case, the `WithViewStore` from `RootScene` doesn't enclose the two scoped views. When `userIsLoggedIn` changes, only the content of its closure is re-run. As the two scoped views are outside of it and there's nothing else in this view to trigger a whole `body` run, these views and the scoped stores are preserved. This root view's body will only run if `store` changes, which is unlikely in this position. You should retrieve the same behavior as before if you move `ChildSceneA/B.View` inside `WithViewStore`.

SwiftUI doesn't invalidate whole subtrees, but only an orbit of views for which it knows or must suppose that the content has changed (because their dependencies have changed, or some `@ObservedObject` emitted `objectWillChange()`). So it furthermore knows that it only needs to invalidate the `WithViewStore` parts of your child views, jumping over the `GroupBox` where you log your redraws.

You can even replace `ChildSceneA/B.View`s by their `body`'s in `RootScene.View` to get a better understanding of the global picture. In this case, you will spot three disjunct `WithViewStore`s, each observing a very narrow piece of state, and each updating iff this piece of state has changed. Everything that is outside a `WithViewStore` has nothing to make it invalidate in the updated toy project you linked.
I hope that makes sense.

**Reply by ghost** [reply]
Thank you @tgrapperon. Yes, it makes perfect sense. I keep forgetting that view builders aren't sequential chunks of code that get executed in order but a way to accumulate view "definitions" that get stored separately and that can be executed separately.
